spinoso-string/src/enc/mod.rs

```diff
@@ -0,0 +1,462 @@
+mod ascii;
+mod binary;
+mod impls;
+mod utf8;
+
+use alloc::vec::Vec;
+use core::ops::Range;
+use core::slice::SliceIndex;
+
+use ascii::AsciiString;
+use binary::BinaryString;
+use utf8::Utf8String;
+
+use crate::codepoints::InvalidCodepointError;
+use crate::encoding::Encoding;
+use crate::iter::{Bytes, IntoIter, Iter, IterMut};
+use crate::ord::OrdError;
+
+#[derive(Clone)]
```

Comment:
Can we impl `Hash`, `PartialEq`, `Eq`, `PartialOrd`, and `Ord` on this enum so we can derive these on `String`?

Something like `Hash` and `PartialEq` in terms of inner byte slice. `PartialOrd` and `Ord` in terms of inner byte slice.

Reply:
Rust seems to not complain about just using the default impl traits here (e.g. `#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]`). From my understanding, this only works since I added the same to the encoded types - which works because `Vec<u8>` support these traits too. Is there any advantages to doing a custom `impl` on these are redirecting to `as_slice()` instead or is the default traits good enough?